### ISSUE TITLE
Enable gzip compression for additional content type(s)

### ISSUE THREAD

**Issue by sunjayBhatia** [projectcontour/contour#4345]
If I wanted to enable compression for another content type, how should I proceed?

_Originally posted by @bourquep in https://github.com/projectcontour/contour/issues/731#issuecomment-1039496574_

**Comment by sunjayBhatia** [comment]
@bourquep started a new issue for this discussion so we can track it correctly

**Comment by sunjayBhatia** [comment]
Currently only gzip and brotli compression are supported in Envoy: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/compression/compression

What sort of content compression are you looking for?

**Comment by bourquep** [comment]
We use Grpc-web (via Envoy via Contour), our payloads are `application/grpc-web+proto`

**Comment by bourquep** [comment]
Just to clarify: I am not looking to enable other _compression_ algorithms, gzip is fine. I need to configure Envoy to gzip-compress another `content-type` besides those listed in #731

**Comment by sunjayBhatia** [comment]
you would have to `applicationgrpc-web+proto` to the defaults from the compressor configuration: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/compressor/v3/compressor.proto#extensions-filters-http-compressor-v3-compressor

contour doesnt modify them here: https://github.com/projectcontour/contour/blob/3498b966d4ed9ef1edb139b8ec8f8b6bddba6c8b/internal/envoy/v3/listener.go#L235

seems reasonable to add it by default, esp. since we default to adding the grpc-web filter as well, so if a client says they accept gzip encoding they get it without having to set anything in contour config

**Comment by bourquep** [comment]
Meanwhile, is there anything I can do to override Envoy's configuration to enable gzip compression for Grpc-web traffic?

**Comment by youngnick** [comment]
Right now, there is no way to override Envoy's config for this at the moment, sorry.

**Comment by sunjayBhatia** [comment]
If you’re open to submitting a PR we would gladly take it 👍🏽

**Comment by bourquep** [comment]
I _might_ give it a shot!

**Comment by bourquep** [comment]
I gave it another try, and I have something that looks in order:

https://github.com/Studyo/contour/commit/79ba90776b88ed201b20864860c37cf0cf718287

But I'm not sure about having to copy-paste the headers in each test though, and I couldn't figure out how to write feature and/or end-to-end tests for it.

I will only be to test it in my cluster after I have upgraded Contour to the latest version, which should be soon.

In other words, I have a commit but not a PR yet. 🤷‍♂️

**Comment by bourquep** [comment]
I have back ported the commit to Contour 1.14.1 (the version we are using in prod), and tested with an updated image.

Unfortunately, it doesn't work - the Grpc-web responses are not compressed. I tried moving the grpcweb filter above the compressor filter, still no luck.

Then I researched a bit more and found this: https://github.com/envoyproxy/envoy/issues/6632

Looks like Envoy isn't able to compress Grpc-web responses after all.

**Comment by youngnick** [comment]
Oh, that sucks, sorry @bourquep. Let's mark this as blocked on Envoy for now then.

**Comment by bourquep** [comment]
There is hope: I upgraded Contour to 1.15.1 (was on 1.14.1), then applied my patch to enable gzip on Grpc-web payloads, and it works!

**Comment by sunjayBhatia** [comment]
Nice! Yeah the Envoy 1.18 release notes mention `compression: extended the compression allow compressing when the content length header is not present. This behavior may be temporarily reverted by setting envoy.reloadable_features.enable_compression_without_content_length_header to false.`: https://www.envoyproxy.io/docs/envoy/latest/version_history/v1.18.0#new-features

**Comment by sunjayBhatia** [comment]
Unfortunately Contour 1.15 is a pretty old release and I'm not sure we will backport the change that far, but if you're willing to submit a PR we can at least enable it in `main`

**Comment by bourquep** [comment]
@sunjayBhatia I've just submitted #4403 for this!
